---
title: "Examine exon to intron ratios"
author: 
- name: Rick Farouni
  affiliation:
  - &cruk Department of Human Genetics, McGill University,  Montreal, Canada
date: '`r format(Sys.Date(), "%Y-%B-%d")`'
output: html_notebook
---


# Preparations

```{r setup}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file(), 
                     fig.width=14,
                     fig.height=14)
set.seed(42)
```


```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(AnnotationHub)
library(biomaRt)
```


### Get exon and intron lengths

```{r}
get_transcriptome <- function(ah_id) {
  ah <- AnnotationHub()
  edb <- ah[[ah_id]]
  seqlevelsStyle(edb) <- "UCSC"
  standard_chroms <- standardChromosomes(edb, species = organism(edb))
  edb <- addFilter(edb, SeqNameFilter(standard_chroms))
  return(edb)
  
}

get_length <- function(x){
  sum(
    width(x))
}

```



```{r}
edb_hg <- get_transcriptome("AH73881")
edb_hg
```
```{r}
columns(edb_hg)
```

```{r}
df_hg <- 
  AnnotationDbi::select(edb_hg,
                        AnnotationDbi::keys(edb_hg,
                                                keytype = "TXID"),
                            keytype = "TXID",
                            columns = c("GENEID", 
                                        "GENENAME",
                                        "GENEBIOTYPE",
                                        "SEQNAME",
                                        "GENESEQSTART",
                                        "GENESEQEND")) 
df_hg
```




```{r}
#exons_by_gene <- exonsBy(edb_hg, by="gene")
exons_by_tx <- exonsBy(edb_hg, by="tx")
introns_by_tx <- intronsByTranscript(edb_hg)
```

```{r}

total_exon_length_by_tx <- lapply(exons_by_tx, get_length)

total_exon_length_by_tx <-
  total_exon_length_by_tx %>%
  unlist() %>%
  enframe( name = "ID", value = "exonic_len")

total_intron_length_by_tx <- lapply(introns_by_tx, get_length)

total_intron_length_by_tx <-
  total_intron_length_by_tx %>%
  unlist() %>%
  enframe( name = "ID", value = "intronic_len")


total_length_by_tx <- full_join(total_exon_length_by_tx,
                                total_intron_length_by_tx, 
                                by="ID")

total_length_by_tx <-
  total_length_by_tx %>%
  dplyr::rename(TXID="ID")


total_length_by_tx <-
  left_join(total_length_by_tx,
            df_hg,
            by="TXID")


total_length_by_tx <-
  total_length_by_tx %>%
  mutate(gene_len= abs(GENESEQEND- GENESEQSTART)+1)
total_length_by_tx
```
## Add gene GC content variable


```{r}
#  searchAttributes(mart, pattern = "length")
#listEnsemblArchives()
mart <- useMart("ensembl", 
                host="http://jul2019.archive.ensembl.org")
mart <- useDataset("hsapiens_gene_ensembl",
                   mart = mart)
bm <- getBM(attributes=c("ensembl_gene_id",
                         'percentage_gene_gc_content'),
            filters = 'ensembl_gene_id',
            values = total_length_by_tx$GENEID,
            mart = mart)
bm <-
  bm %>%
  dplyr::rename(GENEID="ensembl_gene_id")
  
```


```{r}
total_length_by_tx <-
  left_join(total_length_by_tx,
            bm,
            by="GENEID")
total_length_by_tx
```


```{r}
write_csv( total_length_by_tx, "transcripts_hg_ens97_metadata.txt")
```




```{r message=FALSE, warning=FALSE, fig.height=4, fig.width=7}
ggplot(total_length_by_tx %>%
         dplyr::group_by(GENEID) %>%
         dplyr::slice(1),
       aes(percentage_gene_gc_content)) +
  geom_histogram(bins = 170)  
```


```{r, fig.width=20}
ggplot(total_length_by_tx %>%
         dplyr::group_by(GENEID) %>%
         dplyr::slice(1), 
       aes(percentage_gene_gc_content, gene_len))+
  geom_point(size=0.8,
              alpha=0.7) +
  geom_density_2d() +  
   scale_y_log10() +
   theme(legend.position="none")
```






```{r}
total_length_by_tx_protein <-
  total_length_by_tx %>% 
  dplyr::filter(GENEBIOTYPE=="protein_coding")
total_length_by_tx_protein
```

```{r message=FALSE, warning=FALSE, fig.height=4, fig.width=7}
ggplot(total_length_by_tx_protein %>%
         dplyr::group_by(GENEID) %>%
         dplyr::slice(1),
       aes(percentage_gene_gc_content)) +
  geom_histogram(bins = 150)  
```



```{r, fig.width=20}
ggplot(total_length_by_tx_protein%>%
         dplyr::group_by(GENEID) %>%
         dplyr::slice(1), 
       aes(percentage_gene_gc_content, gene_len))+
  geom_point(size=0.8,
              alpha=0.7) +
  geom_density_2d() +  
   scale_y_log10() +
   theme(legend.position="none")
```


```{r, fig.width=20}
ggplot(total_length_by_tx_protein) +
  geom_point( aes(exonic_len, intronic_len, colour=percentage_gene_gc_content),
              size=0.05,
              alpha=0.1)  +  
  scale_x_log10() +
   scale_y_log10() +
  facet_wrap(~SEQNAME)

```

```{r, fig.width=20}
ggplot(total_length_by_tx_protein %>%
         dplyr::filter(SEQNAME=="chr22") )+
  geom_point( aes(exonic_len, intronic_len, colour=gene_len),
              size=0.8,
              alpha=0.7) +
  geom_density_2d(aes(exonic_len, intronic_len)) +  
  scale_x_log10() +
   scale_y_log10() +
   theme(legend.position="none")
```

## Number of transcipts by gene

```{r}
count_tx_by_gene <-
  total_length_by_tx_protein %>%
  count(GENEID, GENENAME) %>%
  arrange(-n)
count_tx_by_gene 
```

Some gene names are not unique

```{r}
duplicated_gene_names <-
  count_tx_by_gene%>%
  dplyr::filter( duplicated(GENENAME ))%>%
  pull(GENENAME)
duplicated_gene_names
```

```{r}
count_tx_by_gene %>%
  dplyr::filter(GENENAME%in% duplicated_gene_names ) %>%
  arrange(GENENAME)
```



```{r message=FALSE, warning=FALSE, fig.height=4, fig.width=7}
ggplot(count_tx_by_gene, aes(n)) +
  geom_histogram(bins = 100)  
```
```{r}
top_genes <- 
  count_tx_by_gene %>%
  top_n(20) %>%
  pull(GENENAME)
top_genes
```


```{r fig.height=8}
ggplot(total_length_by_tx_protein %>% 
         dplyr::filter(GENENAME%in% top_genes))+
  geom_point( aes(exonic_len, intronic_len, colour=gene_len),
              size=0.8,
              alpha=0.7) +
  facet_wrap(~GENENAME)
```


```{r}
ggplot(total_length_by_tx_protein %>% 
         dplyr::filter(GENENAME=="TSC2"))+
  geom_point( aes(exonic_len, intronic_len),
              size=0.8,
              alpha=0.7)
```


```{r}
single_transcript_genes <-
  count_tx_by_gene %>% 
  dplyr::filter(n==1) %>% 
  pull(GENENAME)

```



```{r fig.height=8}
ggplot(total_length_by_tx_protein %>% 
         dplyr::filter(GENENAME%in% single_transcript_genes ))+
  geom_point( aes(exonic_len, intronic_len, colour=gene_len),
              size=0.8,
              alpha=0.7)  +  
  scale_x_log10() +
   scale_y_log10()
```



```{r}
sessionInfo()
```